sqs_workers/queue.py

```diff
@@ -450,16 +452,35 @@ def _should_flush_batch(self) -> bool:
         want to improve that by also ensuring the batch size is small enough.
         """
         max_size = SEND_BATCH_SIZE if self._batch_level > 0 else 1
+
+        # Attempt to flush if we have gotten close to the message size limit
+        if self._est_message_size() > MAX_SQS_MESSAGE_SIZE:
+            return True
+
         return len(self._batched_messages) >= max_size
 
+    def _est_message_size(self) -> int:
+        """
+        Return an estimate of the size (in bytes) of the combined message we want to send.
+
+        We add an extra 1K to account for any extra headers.
+        """
+        return len(json.dumps(self._batched_messages).encode("utf-8")) + 1024
+
     def _flush_batch_if_needed(self) -> None:
         queue = self.get_queue()
 
         # There should be at most 1 batch to send. But just in case, prepare to
         # send more than that.
         while self._should_flush_batch():
-            msgs = self._batched_messages[:SEND_BATCH_SIZE]
-            self._batched_messages = self._batched_messages[SEND_BATCH_SIZE:]
+            send_batch_size = SEND_BATCH_SIZE
+
+            if self._est_message_size() > MAX_SQS_MESSAGE_SIZE:
+                # If we have batch of large messages, try to send half of it at a time
+                send_batch_size = SEND_BATCH_SIZE / 2
```

Comment:
What about:

```python
while send_batch_size > 1 and self._est_message_size(self._batched_messages[:send_batch_size]) > MAX_SQS_MESSAGE_SIZE:
	send_batch_size -= 1
```

This would be more correct than halving and hoping for the best. 🤔 

It would require tweaking `_estimate_message_size` to receive `batched_messages` as an argument, too.

Reply:
It would be good to have a test for this, too!

Reply:
Good call @goncalossilva. Factoring out the `_estimate_message_size` code removed the need to guess and hope for the best. :rocket: